lib/charms/observability_libs/v0/juju_topology.py

```diff
@@ -0,0 +1,230 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+"""## Overview.
+
+This document explains how to use the JujuTopology class to
+create and consume topology information in a consistent manner.
+
+The goal of the Juju topology is to uniquely identify a piece
+of software running across any of your Juju-managed deployments.
+This is achieved by combining the following four elements:
+
+- Model name
+- Model UUID
+- Application Name
+- Unit Identifier
+
+and in some cases, the charm name.
+
+For a more in-depth description of the concept, as well as a
+walk-through of it's use-case in observability, see
+[this blog post](https://juju.is/blog/model-driven-observability-part-2-juju-topology-metrics)
+on the Juju blog.
+
+## Library Usage
+
+This library may be used to create and consume JujuTopology objects.
+The JujuTopology class provides three ways to create instances:
+
+### Using the `from_charm` method
+
+Enables instantiation by supplying the charm as an argument. When
+creating topology objects for the current charm, this is the recommended
+approach.
+
+```python
+topology = JujuTopology.from_charm(self)
+```
+
+### Using the `from_dict` method
+
+Allows for instantion using a dictionary of relation data, like the
+`scrape_metadata` from Prometheus or the labels of an alert rule. When
+creating topology objects for remote charms, this is the recommended
+approach.
+
+```python
+scrape_metadata = json.loads(relation.data[relation.app].get("scrape_metadata", "{}"))
+topology = JujuTopology.from_dict(scrape_metadata)
+```
+
+### Using the class constructor
+
+Enables instantiation using whatever values you want. While this
+is useful in some very specific cases, this should not be considered
+the least favored approach.
+
+```python
+topology = JujuTopology(
+    model="some-juju-model",
+    model_uuid="00000000-0000-0000-0000-000000000001",
+    application="fancy-juju-application",
+    unit="fancy-juju-application/0",
+    charm_name="fancy-juju-application-k8s",
+)
+```
+
+"""
+
+from collections import OrderedDict
+from typing import Dict, List, Optional
+
+# The unique Charmhub library identifier, never change it
+LIBID = "bced1658f20f49d28b88f61f83c2d232"
+
+LIBAPI = 0
+LIBPATCH = 1
+
+
+class JujuTopology:
```

Comment:
> I imagine that for a juju dev "juju topology" would mean something very different.

If it does, we have an issue. The term was, to my knowledge, coined in observability by @mmanciop, but it should be valid regardless of subcontext as long as we still talk about Juju deployments.

Reply:
> The goal is less to classify, and more to identify

Totally agree, that's why I don't like `Taxonomy` for this.
But JujuTopolgy does not translate in my mind into "identifying app/unit". Nor does `CharmTopology`.
I.e. "network topology" is not analogous to "juju topology" the way we currently use it.

Reply:
It is definitely possible that it's one of those things which comes about when you have to manage an environment which is unnecessarily large, and naming things after planets/mythology/whatever doesn't work anymore. 

"I got an alert. How do I can I tell from the information I already have where it's located, what it does, and a guess at how many of these there are so I know what the impact is going to be?"

It's kinda "topology as seen by Juju", since we're not adding the controller name (which isn't exposed in any meaningful way in OF or in Juju anyway -- see [here](https://github.com/juju/juju/blob/ebaa903b73f02ff51a7386274b8de1cd292abd79/worker/uniter/runner/context/context.go#L1067) for a more or less list of what could conceivably be added), but it's still in the sense of "this is how things are arranged."

The name is sort of aiming for understanding from ops/admins rather than devs.

Reply:
Taxonomy implies "classification" of charms, Topology implies geometric arrangement. I think "Juju Topology" is more appropriate since the topology information is used to provide information of the location (geometric property) i.e. source of the metrics, etc.

Reply:
I'm killing this conversation. We're not renaming it. 😅